feat: add optional name for activity descriptions

Anyitechs opened the pull request:
This PR closes issue #127  and the implementation follows the conversation on the issue thread.

carlaKC commented on sim-lib/src/lib.rs:
```diff
@@ -573,19 +578,25 @@ impl Simulation {
 
+                let activity_name = match &description.activity_name {
+                    Some(name) => name.clone(),
+                    None => index.to_string(),
```
perhaps "index {}" so that users understand where the number comes from

Anyitechs replied:
Sure! I'll fix that alongside the other changes and update the PR.

carlaKC commented on sim-lib/src/lib.rs:
```diff
@@ -790,13 +803,14 @@ async fn consume_events(
     listener: Listener,
 ) {
-    log::info!("Started activity producer for {source}: {node_generator}.");
+    log::info!("Started {activity_name} => activity producer for {source}: {node_generator}.");
```
Let's have the `activity_name` as a prefix for all te logs in `produce_events` so that they're easily reference-able. Extra points if it's propagated through to other activity-relevant logging.

carlaKC commented on sim-lib/src/lib.rs:
```diff
@@ -637,6 +648,7 @@ impl Simulation {
             generators.push(ExecutorKit {
                 source_info: node_info.clone(),
                 network_generator: network_generator.clone(),
+                activity_name: "Random activity".to_string(),
```
This doesn't really add much, I'd just have no name for randomly generated activity.

carlaKC reviewed (commented):
Conceptually looks good, a few comments. Feel free to pull out of draft once addressed.

Anyitechs commented:
@carlaKC  I made the requested changes. Let me know if this looks good now. Thank you.

enigbe reviewed (changes requested):
@Anyitechs Good job on the PR. Here are a few comments I have:

1. To make the new `activity_name` field even clearer for users, I'd recommend adding some documentation to the README. This could explain its optional nature and how it's handled in the code.

2. I noticed that the `activity_name` or `index` is used in logs related to activities and at most the simulation events (when they're sent to the events consumer). I'm curious about why we'd have a name/index prefix that doesn't extend to the simulation outputs and results as well. Shouldn't there be consistency across board? I don't think it will be easy to relate `activity -> event -> output -> result` logs if only the activities and the events have the prefix. I know this is out of the scope of the issue but happy to get your (as well as @carlaKC's) thoughts on this.

Anyitechs commented:
> @Anyitechs Good job on the PR. Here are a few comments I have:

Thank you.

> 
> 1. To make the new `activity_name` field even clearer for users, I'd recommend adding some documentation to the README. This could explain its optional nature and how it's handled in the code.

Sure! I'll update the README in a seperate PR following this.

> 2. I noticed that the `activity_name` or `index` is used in logs related to activities and at most the simulation events (when they're sent to the events consumer). I'm curious about why we'd have a name/index prefix that doesn't extend to the simulation outputs and results as well. Shouldn't there be consistency across board? I don't think it will be easy to relate `activity -> event -> output -> result` logs if only the activities and the events have the prefix. I know this is out of the scope of the issue but happy to get your (as well as @carlaKC's) thoughts on this.

I agree @enigbe . I guess that's what @carlaKC referred to as "activity-related" logging above. I'll work on that and update the PR.

carlaKC commented:
> Sure! I'll update the README in a seperate PR following this.

Please update readme in the same PR, it's easy for things to get out of sync/be forgotten otherwise.

carlaKC reviewed (commented):
Shaping up! 

I think that this still needs better handling for random activity - what we have right now will print odd looking lines.

Anyitechs commented on README.md:
```diff
@@ -136,8 +136,11 @@ of the random activity that is generated:
 and destination nodes, you can provide `activity` descriptions of the 
-source, destination, frequency and amount for payments that you'd like 
-to execute. Note that `source` nodes *must* be contained in `nodes`, 
+source, destination, frequency, amount for payments that you'd like 
+to execute and an optional name for each `activity` description that is used
+as the logging prefix to relate logs to their corresponding activity description 
+(The index of each `activity` description will be used if a name is not 
+provided for the `activity` description). Note that `source` nodes *must* be contained in `nodes`, 
```
Sure, I'll update that. Thank you.

Anyitechs commented on sim-lib/src/lib.rs:
```diff
@@ -819,32 +838,32 @@ async fn produce_events<N: DestinationGenerator + ?Sized, A: PaymentGenerator +
                     Ok(amt) => {
                         if amt == 0 {
-                            log::debug!("Skipping zero amount payment for {source} -> {destination}.");
+                            log::debug!("{activity_name} activity => Skipping zero amount payment for {source} -> {destination}.");
```
Ok. I'll replace that. 

And if a name isn't set, it should print the `index` of the `activity`. But I'll confirm this is true for this as well.

carlaKC replied:
We don't set an `activity_name` for randomly generated activity (above, line 776) right?

Anyitechs replied:
Yes, we don't set an `activity_name` for randomly generated activity. I see your point on this logging ` activity` if a name isn't set for it.

How best do you think we can handle this? I had a name set for it before but you earlier mentioned we should take it off.

carlaKC replied:
By only logging the prefix if we've got a value there?

`if self.activity_name != "" {  format!("{} activity:", self.activity_name) } else {"".to_string()}`

Could inline, pulled out into a impl of `Payment`, or leave `activity_name` as an option so that we can use some of the `uwrap_or` functionality.

Anyitechs replied:
Thank you for the clarification. I've addressed this and updated the PR.

Anyitechs commented on sim-cli/src/main.rs:
```diff
@@ -187,6 +187,7 @@ async fn main() -> anyhow::Result<()> {
             destination,
             interval_secs: act.interval_secs,
             amount_msat: act.amount_msat,
+            activity_name: act.activity_name,
```
Yes, I agree. How best do you think we can make them unique? Does appending the event name to the activity name work?

carlaKC reviewed (commented):
Needs a rebase on master!

carlaKC commented on sim-cli/src/main.rs:
```diff
@@ -182,11 +182,20 @@ async fn main() -> anyhow::Result<()> {
+        for validated_activity in &validated_activities {
+            if validated_activity.activity_name == act.activity_name {
+                anyhow::bail!(LightningError::ValidationError(
+                    "Duplicate activity name is not allowed".to_string(),
```
It would be helpful to provide the user with the name so they know what to fix?

carlaKC commented on sim-lib/src/lib.rs:
```diff
@@ -935,19 +960,32 @@ impl PaymentResultLogger {
+        if self.activity_name.is_empty() {
+            write!(
+                f,
+                "Processed {} payments sending {} msat total with {:.2}% success rate.",
```
Surely we can do this with fewer LOC?

carlaKC commented on sim-cli/src/main.rs:
```diff
@@ -182,11 +182,20 @@ async fn main() -> anyhow::Result<()> {
             },
         };
 
+        for validated_activity in &validated_activities {
```
Build a hashmap like we do for alias and pubkey rather than iterating through every activity every time

carlaKC commented on README.md:
```diff
@@ -84,17 +92,18 @@ Whereas for CLN nodes, the following information is required:
-provide a set of nodes and the simulator will generate activity based 
-on the topology of the underlying graph. Note that payments will only 
-be sent between the `nodes` that are provided so that liquidity does 
+To run the simulator with random activity generation, you just need to
```
Please pull wholesale formatting changes out into their own commit

carlaKC replied:
This still hasn't been addressed?

If you want to run all these formatting changes on the README, please run them first and then make your changes.

Anyitechs commented:
Thank you for the detailed review @carlaKC. I've addressed all the feedbacks, rebased, and updated the PR as well.

Let me know if there's anything else I need to look at.

carlaKC commented on sim-lib/src/lib.rs:
```diff
@@ -1127,12 +1122,17 @@ impl PaymentResultLogger {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         let total_payments = self.success_payment + self.failed_payment;
+
+        // if self.activity_name
```
Incomplete comment and TODO left in this commit?

carlaKC replied:
Also unaddressed?

carlaKC replied:
Then the [fixup should be squashed](https://jordanelver.co.uk/blog/2020/06/04/fixing-commits-with-git-commit-fixup-and-git-rebase-autosquash/) as requested.

Anyitechs replied:
Ok. Thank you.

carlaKC reviewed (commented):
Looks like the functionality is mostly here (though I think we still need to clean up the logging for random activity). 

On a higher level, the commit structure in this PR needs some work. Most of the changes can be squashed down to a single commit, and formatting changes / docs / unrelated fixups left in their own commit.

Anyitechs commented:
> Looks like the functionality is mostly here (though I think we still need to clean up the logging for random activity).
> 
> On a higher level, the commit structure in this PR needs some work. Most of the changes can be squashed down to a single commit, and formatting changes / docs / unrelated fixups left in their own commit.

Thank you. I've addressed the feedbacks and neatened up the commit structure.

carlaKC commented on README.md:
WDYM? This commit also has all of the rust changes in it?

Anyitechs replied:
I'm sorry I was referring to this [commit](https://github.com/bitcoin-dev-project/sim-ln/pull/170/commits/19f7162aaa229e15185fa4de8aff1c5810987d1f) and I don't think it has all of the rust changes in it. But I'll address the format issue in the first commit.

carlaKC replied:
Please see [suggested structure](https://github.com/bitcoin-dev-project/sim-ln/pull/170#issuecomment-2073242761) and [contribution guidelines](https://github.com/bitcoin-dev-project/sim-ln/blob/main/CONTRIBUTING.md#commit-history).

carlaKC commented on sim-lib/src/lib.rs:
```diff
@@ -358,6 +358,16 @@ struct Payment {
     dispatch_time: SystemTime,
 }
 
+impl Payment {
```
Can be squashed with the original rust changes?

Anyitechs commented:
> > Thank you. I've addressed the feedbacks and neatened up the commit structure.
> 
> Have you? The first commit is still a mix of formatting and logical changes, and has a dangling TODO comment.
> 
> To me, the logical commit structure would be:
> 
> * Formatting changes to the README (or remove all of those)
> * Commit with actual rust changes
> * Commit with documentation updates for features
> 
> Right now everything is intermingled in [the first commit](https://github.com/bitcoin-dev-project/sim-ln/pull/170/commits/1f2920772a2e33e8215c3a13f447fe49a82586c0) - formatting changes to the readme, and rust code.

Thank you @carlaKC for taking the time to review this and providing feedbacks as well. I really appreciate it.

I've worked on the commit structure and currently squashed it into two commits:

- The actual code implementation, and
- The documentation update of the feature

Let me know if I missed anything. Thank you.

carlaKC reviewed (changes requested):
Thanks for updating the commit ordering! 

Please try to test changes with a few standard inputs before you push/request review - there's a panic in this change. 

The various logging issues that I've pointed out also become more apparent when you spend some time running the project and getting familiar. I know that these seem like frustrating nits, but when we're looking at a change like this which really is just aesthetic we really do need to get things right to justify changing the code at all.

carlaKC commented:
If it's useful to getting this in, we can also just re-add `random activity` as a name, but perhaps with an index. I don't like it but it's not really I hill I'm willing to die on - hopefully that simplifies all the logging.

carlaKC commented:
Needs rebase!

Anyitechs commented:
> Ping @Anyitechs any updated here?

Yes, will update the PR by EOD.

Anyitechs commented:
> > Ping @Anyitechs any updated here?
> 
> Yes, will update the PR by EOD.

I just updated the PR now.

carlaKC commented on sim-lib/src/lib.rs:
```diff
@@ -1116,30 +1150,40 @@ struct PaymentResultLogger {
+            let total_payments = activity_result.success_payment + activity_result.failed_payment;
+            writeln!(
+                f,
+                "{} Processed {} payments sending {} msat total with {:.2}% success rate.",
```
The intent behind [this comment](https://github.com/bitcoin-dev-project/sim-ln/pull/170#discussion_r1578293858) was to remove the names from `PaymentResultLogger` completely - we don't want to print these results on a per-activity basis, it'll be too spammy where the whole point here is to give a very high level aggregate.

I believe we can also remove `activity_name` from `Payment` completely if we don't have this logging?

carlaKC commented on sim-lib/src/lib.rs:
```diff
@@ -352,6 +358,16 @@ struct Payment {
 }
 
+impl Payment {
+    fn activity_prefix(&self) -> String {
```
nit: add a comment explaining usage here / that this should be added before a log with no space

carlaKC commented on sim-lib/src/lib.rs:
```diff
@@ -352,6 +358,16 @@ struct Payment {
+impl Payment {
+    fn activity_prefix(&self) -> String {
+        if !self.activity_name.is_empty() {
+            format!("{} activity:", self.activity_name)
```
nit: should have a space after `:` otherwise it reads `X activity:Payment` when we use it below

carlaKC commented on sim-lib/src/lib.rs:
```diff
@@ -788,10 +812,11 @@ impl Simulation {
             generators.push(ExecutorKit {
                 source_info: node_info.clone(),
                 network_generator: network_generator.clone(),
+                activity_name: format!("Index {index} random"),
```
nit: `Random index {}`

carlaKC commented on sim-lib/src/lib.rs:
```diff
@@ -788,10 +812,11 @@ impl Simulation {
         );
 
-        for (node_info, capacity) in active_nodes.values() {
+        for (index, (node_info, capacity)) in active_nodes.values().enumerate() {
```
nit: just use i, more idiomatic

carlaKC reviewed (changes requested):
Can be rebased on master to fix unrelated build failure. 

I think that this change is high impact enough if we just have logging in `produce_events`, so final comment is to remove this entirely from `PaymentResultLogger`.

carlaKC reviewed (changes requested):
There are some superfluous changes leftover from previous iterations of this PR, and `activity_name` isn't being used in logs in `produce_events` in several places.

carlaKC commented:
Closing due to inactivity.
